[`pylint`] Re-implement `unreachable` (`PLW0101`)

augustelalande opened the pull request:
## Summary

This PR re-introduces the control-flow graph implementation which was first introduced in #5384, and then removed in #9463 due to not being feature complete. Mainly, it lacked the ability to process `try`-`except` blocks, along with some more minor bugs.

Closes #8958 and #8959 and #14881.

## Overview of Changes

I will now highlight the major changes implemented in this PR, in order of implementation.

1. Introduced a post-processing step in loop handling to find any `continue` or `break` statements within the loop body and redirect them appropriately.
2. Introduced a loop-continue block which is always placed at the end of loop blocks, and ensures proper looping regardless of the internal logic of the block. This resolves #8958.
3. Implemented `try` processing with the following logic (resolves #8959):
    1. In the example below the cfg first encounters a conditional `ExceptionRaised` forking if an exception was (or will be) raised in the try block. This is not possible to know (except for trivial cases) so we assume both paths can be taken unconditionally.
    2. Going down the `try` path the cfg goes `try`->`else`->`finally` unconditionally.
    3. Going down the `except` path the cfg will meet several conditional `ExceptionCaught` which fork depending on the nature of the exception caught. Again there's no way to know which exceptions may be raised so both paths are assumed to be taken unconditionally.
    4. If none of the exception blocks catch the exception then the cfg terminates by raising a new exception.
    5. A post-processing step is also implemented to redirect any `raises` or `returns` within the blocks appropriately.
```python
def func():
    try:
        print("try")
    except Exception:
        print("Exception")
    except OtherException as e:
        print("OtherException")
    else:
        print("else")
    finally:
        print("finally")
```
```mermaid
flowchart TD
  start(("Start"))
  return(("End"))
  block0[["`*(empty)*`"]]
  block1["print(#quot;finally#quot;)\n"]
  block2["print(#quot;else#quot;)\n"]
  block3["print(#quot;try#quot;)\n"]
  block4[["Exception raised"]]
  block5["print(#quot;OtherException#quot;)\n"]
  block6["try:
        print(#quot;try#quot;)
    except Exception:
        print(#quot;Exception#quot;)
    except OtherException as e:
        print(#quot;OtherException#quot;)
    else:
        print(#quot;else#quot;)
    finally:
        print(#quot;finally#quot;)\n"]
  block7["print(#quot;Exception#quot;)\n"]
  block8["try:
        print(#quot;try#quot;)
    except Exception:
        print(#quot;Exception#quot;)
    except OtherException as e:
        print(#quot;OtherException#quot;)
    else:
        print(#quot;else#quot;)
    finally:
        print(#quot;finally#quot;)\n"]
  block9["try:
        print(#quot;try#quot;)
    except Exception:
        print(#quot;Exception#quot;)
    except OtherException as e:
        print(#quot;OtherException#quot;)
    else:
        print(#quot;else#quot;)
    finally:
        print(#quot;finally#quot;)\n"]

  start --> block9
  block9 -- "Exception raised" --> block8
  block9 -- "else" --> block3
  block8 -- "Exception" --> block7
  block8 -- "else" --> block6
  block7 --> block1
  block6 -- "OtherException" --> block5
  block6 -- "else" --> block4
  block5 --> block1
  block4 --> return
  block3 --> block2
  block2 --> block1
  block1 --> block0
  block0 --> return
``` 
6. Implemented `with` processing with the following logic:
    1. `with` statements have no conditional execution (apart from the hidden logic handling the enter and exit), so the block is assumed to execute unconditionally.
    2. The one exception is that exceptions raised within the block may result in control flow resuming at the end of the block. Since it is not possible know if an exception will be raised, or if it will be handled by the context manager, we assume that execution always continues after `with` blocks even if the blocks contain `raise` or `return` statements. This is handled in a post-processing step.

## Test Plan

Additional test fixtures and control-flow fixtures were added.

charliermarsh commented:
Nice, I'm a fan of this.

augustelalande commented:
No promises, just started taking a look at it.

augustelalande commented:
Can you give me some background on why this was never activated? Was there just too many false positives?

charliermarsh commented:
No, I think it's fairly reliable, but I believe `try`-`except` handling wasn't quite finished: https://github.com/astral-sh/ruff/issues/8959. There's also at least one bug to fix: https://github.com/astral-sh/ruff/issues/8958.

augustelalande commented:
Resolve #8958

augustelalande commented:
This is probably not completely ready for merging, but considering the size it's probably worth getting some feedback now.

augustelalande commented:
@charliermarsh when you get the time please have a look at this

MichaReiser commented:
Sorry for the emberassing long wait. I make this a priority for next week. 

The best way to review this change is probably to compare all changes starting *after* the "Revert" commit. If you have any other review recommendations, let me know :)

augustelalande commented:
@MichaReiser no worries. Ya that's a good strategy, but also take a look at my PR summary at the top, I tried to give an outline of the changes I made.

MichaReiser commented:
I rebased the commit onto main

MichaReiser commented:
> [ibis-project/ibis](https://github.com/ibis-project/ibis) (error)

I did attach a debugger and it seems that it overflows in `post_process_try` when analysing `"ibis/ibis/backends/pyspark/__init__.py"`

MichaReiser commented on crates/ruff_linter/src/rules/ruff/rules/snapshots/ruff_linter__rules__ruff__rules__unreachable__tests__try.py.md.snap:
````diff
@@ -0,0 +1,494 @@
+
+## Function 10
+### Source
+```python
+def func():
+    try:
+        assert False
+        print("unreachable")
+    finally:
+        print("reached")
+```
````
We should add some tests where we raise exceptions inside blocks other than `try`. E.g.

```python
try:
    assert False
except ex:
    raise ex

finally:
    raise Exception("other")
```

Ideally we combine this with a nested `try` block.

augustelalande replied:
I added a big one, and there were issues, so good catch

MichaReiser commented on crates/ruff_linter/src/rules/ruff/rules/snapshots/ruff_linter__rules__ruff__rules__unreachable__tests__for.py.md.snap:
````diff
@@ -0,0 +1,405 @@
+### Source
+```python
+def func():
+    for i in range(3):
+        if i == 2:
+            assert i is not None
+            break
+    else:
+        raise Exception()
+    x = 0
````
Let's add an example with a nested loop where both the outer and inner loop use break/continue.

augustelalande commented on crates/ruff_linter/src/checkers/ast/analyze/statement.rs:
```diff
@@ -370,6 +370,11 @@ pub(crate) fn statement(stmt: &Stmt, checker: &mut Checker) {
             if checker.enabled(Rule::AsyncFunctionWithTimeout) {
                 flake8_async::rules::async_function_with_timeout(checker, function_def);
             }
+            if checker.enabled(Rule::UnreachableCode) {
+                checker
+                    .diagnostics
+                    .extend(ruff::rules::unreachable::in_function(name, body));
+            }
```
Well lambda body doesn't make sense since it can only a single statement. But module body can definitely be added.

augustelalande commented on crates/ruff_linter/src/rules/ruff/rules/unreachable.rs:
```diff
@@ -0,0 +1,1354 @@
+            _ => None,
+        },
+        Condition::Iterator(_) => None,
+        Condition::Match { .. } => None,
```
Certainly

augustelalande commented on crates/ruff_linter/src/rules/ruff/rules/unreachable.rs:
```diff
@@ -0,0 +1,1354 @@
+    /// `fixup_index` from `from` to `to`.
+    ///
+    /// This doesn't change the target if it's `NextBlock::Terminate`.
+    fn change_next_block(
```
removed

augustelalande commented on crates/ruff_linter/src/rules/ruff/rules/unreachable.rs:
```diff
@@ -0,0 +1,1354 @@
+    // Exception was raised and caught by `except` clause
+    ExceptionCaught(&'stmt Expr),
+    // Exception was raised in a `try` block
+    ExceptionRaised,
```
Changed to `Except` and `MayberRaised`

augustelalande commented on crates/ruff_linter/src/rules/ruff/rules/unreachable.rs:
```diff
@@ -0,0 +1,1354 @@
+    let after_block = blocks.maybe_next_block_index(after, || true);
+    let last_orelse_statement = blocks.append_blocks_if_not_empty(orelse, after_block);
+
+    let loop_continue_index = blocks.blocks.push(BasicBlock::LOOP_CONTINUE.clone());
```
`continue` can and does link to `block 0`, the point of the comment is that this is "odd" since it implies that both `i = 0` and `while True` are executed after each loop, which isn't true but isn't important for the purpose of determining reachability. This behaviour can be changed easily if desired so that `i = 0` and `while True` are seperate blocks.

augustelalande commented on crates/ruff_linter/src/rules/ruff/rules/unreachable.rs:
```diff
@@ -0,0 +1,1354 @@
+        orelse: BlockIndex,
+        /// Exit block. None indicates Terminate.
+        exit: Option<BlockIndex>,
+    },
```
I will improve the documentation, but exit serves a different purpose, not for determining control flow, but rather to facilitate the post_processing steps.

augustelalande commented:
@MichaReiser Ok I think I've addressed all your comments, time for second round of review.

I did have one issue I couldn't resolve. You can see it in one of the test fixtures, but essentially in the following
```python
def if_elif_always_false():
    if False:
        return "unreachable"
    elif False:
        return "also unreachable"
    return "reachable"
```
the two unreachable lines get combined as a single violation
```
RUF014.py:21:9: RUF014 Unreachable code in `if_elif_always_false`
   |
19 |   def if_elif_always_false():
20 |       if False:
21 |           return "unreachable"
   |  _________^
22 | |     elif False:
23 | |         return "also unreachable"
   | |_________________________________^ RUF014
24 |       return "reachable"
```
this is because I combine blocks which follow each other if there are no reachable blocks in between, and since the elif doesn't make its own block, I couldn't figure out how to split them.

augustelalande commented:
Actually I may be able to extract the range of the clauses from the `ElifElseClause`, I will try tomorrow

MichaReiser commented:
I plan to look at this PR again but I don't know yet when I have the time to do so.

MichaReiser commented:
There's a good chance that this rule will become obsolete with Re Knot, but integrating Red Knot back into Ruff is also still far out. I don't mind keeping this PR open for now in the hope that I'll soon have time to review it (I'm so sorry).

dylwil3 commented:
@augustelalande I'm a big fan of this PR! I think it's a great jumping off point for Ruff to begin learning control flow more generally. I'd love to help get this merged, so I'm gonna help with the review process.

Before I dive in to the specifics, two more general questions regarding the implementation:

1. I would think that the beginning of a loop (both for and while) would be its own block, so that the body of the loop can point back to it. As you indicate in a comment, that's not the current behavior. I'm curious what the downside is of adopting what feels like the more intuitive approach?
2. You build the control flow graph starting from the bottom, which is nice because `NextBlock` always knows where to point, but it adds this post-processing step and it is reversed (I think) from the direction of the "unreachable" analysis. An alternative approach would be to build the graph in the forward direction, pointing to empty blocks that are later populated, and performing `mark_unreachable` as you go. Is there a hidden downside to that approach?

Thanks again for all your hard work, and sorry for the long, drawn out review!

augustelalande commented:
@dylwil3 so as I stated in the opening summary, the starting point for this PR is #5384 which already did the bulk of the implementation, and had already been merged into main and then later removed. As such a lot of the design decisions were made by the original author, and I only made changes where I felt they were necessary.

Secondly, even if some of those decisions seem incorrect in retrospect, I'm not prepared at the moment to dedicate much more work to this PR, especially since I think red-knot will eventually implement its own CFG which will supersede this. Of course if someone else wants to take on that works I won't object. So please keep that in mind.

1. I fully agree with this. Having a dedicated block for a loop start is a much better approach. So if this becomes the sticking point preventing a merge I'll be happy to change it.

2. As you stated the main benefit here is to always have the NextBlock available. I don't think the various post-processing steps add much cost in terms of performance, although they may be undesirable in terms of readability. In any case, switching to forward approach would essentially require a full rewrite so I'm unlikely to commit to it.

dylwil3 commented:
> As such a lot of the design decisions were made by the original author, and I only made changes where I felt they were necessary.

I suspected that might be the case - makes sense!

>  Having a dedicated block for a loop start is a much better approach.

If this is not too big of a change, let's do this then

> In any case, switching to a forward approach would essentially require a full rewrite so I'm unlikely to commit to it.

I don't think we need to do this - I was just curious if I was missing anything here.

> Secondly, even if some of those decisions seem incorrect in retrospect, I'm not prepared at the moment to dedicate much more work to this PR

I totally understand! I don't want to let the perfect be the enemy of the good here. I'm hoping we can get this merged in without too much extra work, but please let me know if you'd like me to finish it off or if I can help in any way!

dylwil3 commented:
@augustelalande Upon further reflection, I've decided to merge this in as-is. The various remaining tweaks do not really affect the core functionality of this rule, and could be tackled in follow-up PRs.

This is an outstanding contribution to Ruff, and your hard work deserves to be part of the codebase!

Thank you once again for your patience and persistence!

augustelalande commented:
Thanks @dylwil3
